### PR TITLE
Add value ranges for GeoTile aggregation metrics metrics in the meta layer

### PR DESCRIPTION
**Pull request by iverase**
In order to support automatic styling, the consumers of vector tiles need to know the range of values added to the aggs layer. Therefore this PR adds this values ranges on the meta later. Values are computed using min / max pipeline aggregations.

So for example, there is the document counts range in the form:
```
_count.min
_count.max
```
In addition, for any metric aggregation there will be two entries:
```
aggs.<aggName>.min
aggs.<aggName>.max
```

### CODE REVIEW

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/AbstractVectorTileSearchAction.java [hunk 1]
```diff
@@ -226,9 +234,31 @@ public void setSize(int size) {
         }
 
         public void setAggBuilder(AggregatorFactories.Builder aggBuilder) {
-            // TODO: validation
+            for (AggregationBuilder aggregation : aggBuilder.getAggregatorFactories()) {
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/AbstractVectorTileSearchAction.java:237]: Let's wait with this. I think I have some idea on how to make this more generic.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/VectorTileGeometryBuilder.java [hunk 2]
```diff
@@ -57,10 +57,10 @@ public void box(VectorTile.Tile.Feature.Builder featureBuilder, double minLon, d
     }
 
     private int lat(double lat) {
-        return (int) (pointYScale * (VectorTileUtils.latToSphericalMercator(lat) - rectangle.getMinY())) + extent;
+        return (int) Math.round(pointYScale * (VectorTileUtils.latToSphericalMercator(lat) - rectangle.getMinY())) + extent;
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/VectorTileGeometryBuilder.java:60]: We should still probably get this in.

**Review by imotov — APPROVED** [review]
Let's get the bug-fix in, but let's wait with handling aggregations here. I think I have a more generic solution to this one.

**Comment by iverase** [comment]
Thanks @imotov , I merged your changes into this PR, and the output of the pipeline aggregations are now added to the meta layer using your approach (very cool)

**Comment by imotov** [comment]
Nice!

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/AbstractVectorTileSearchAction.java (hunk 1 above)
Comment (imotov) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/AbstractVectorTileSearchAction.java:237]: I still wonder if 1) this should responsibility of the callers and 2) if we want to add these pipeline aggs, we need to throw an exception if we don't recognize the agg

Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/AbstractVectorTileSearchAction.java:237]: My hope is that a user of this API does not need to be an expert on Elasticsearch so we can provide this information from them. I am hoping to only support a subset of the aggregations, the ones that make sense.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/RestVectorTileAction.java [hunk 3]
```diff
@@ -203,23 +214,9 @@ private static SearchRequestBuilder searchBuilder(SearchRequestBuilder searchReq
...
-                    case SumAggregationBuilder.NAME:
-                    case CardinalityAggregationBuilder.NAME:
-                        final NumericMetricsAggregation.SingleValue metric = (NumericMetricsAggregation.SingleValue) aggregation;
-                        addPropertyToFeature(featureBuilder, layerProps, "aggs." + aggregation.getName(), metric.value());
-                        break;
-                    default:
-                        // top term and percentile should be supported
-                        throw new IllegalArgumentException("Unknown feature type [" + type + "]");
-                }
+                addToXContentToFeature(featureBuilder, layerProps, aggregation);
```
Comment (iverase) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/RestVectorTileAction.java:219]: Let's add the result of aggregations from the Content as well

**Comment by iverase** [comment]
I moved the methods that add properties to the layer to a hyper class and rename some of the methods. @imotov should be push this PR?

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/AbstractVectorTileSearchAction.java [hunk 4]
```diff
@@ -226,9 +234,31 @@ public void setSize(int size) {
...
+                    case SumAggregationBuilder.NAME:
+                    case CardinalityAggregationBuilder.NAME:
+                        break;
+                    default:
+                        // top term and percentile should be supported
+                        throw new IllegalArgumentException("Unsupported aggregation of type [" + type + "]");
+                }
+            }
+            for (PipelineAggregationBuilder aggregation : aggBuilder.getPipelineAggregatorFactories()) {
+                // should not have pipeline aggregations
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/vectortile/AbstractVectorTileSearchAction.java:252]: Why?

**Review by imotov — APPROVED** [review]
Let's get this in, but I would love to iterate on it a bit more next. Something bothers me about this case statement with hardcoded list of 5 aggregations. I feel like it really can be any NumericMetricsAggregator.SingleValue derived agg there.

**Comment by iverase** [comment]
> I feel like it really can be any NumericMetricsAggregator.SingleValue derived agg there.

Yes but I think there is nothing in the aggregation builder that tells you the type of aggregation?

**Comment by imotov** [comment]
Indeed, we need to figure this part out.

**Pull request merged**